src/status_im/core.cljs

```diff
@@ -47,7 +48,13 @@
 (defn init
   []
   (navigation/init)
-  (native-module/init #(re-frame/dispatch [:signals/signal-received %]))
+  (native-module/init
+   (fn [raw-event]
+     ;; In some cases, an event should not be handled until after the GUI has
+     ;; been updated, i.e., after the next Reagent animation frame. In such a
+     ;; case, the event should be dispatched with :flush-dom metadata. For more
+     ;; details, check re-frame's source code and documentation.
+     (re-frame/dispatch ^:flush-dom [:signals/signal-received raw-event])))
```

Comment:
So, is this a way to schedule this event until the end of the frame?

Is this the same as wrapping the event in `js/requestAnimationFrame`? 🤔

If so, I thought reagent (an so re-frame) already executed their tasks at the end of the frame by default.

Reply:
The comment I added is a copy & paste from the documentation. I'm taking re-frame's word for this, check it out from the docs https://github.com/day8/re-frame/blob/09e2d7132c479aa43f2a64164e54e42bf8511902/docs/Solve-the-CPU-hog-problem.md#L195:

```
In these kinds of cases, where you are only going to give the UI 
**one chance to update** (not a repeated chance every few milliseconds), 
then you had better be sure the DOM is fully synced. 

To do this, you put metadata on the event being dispatched:

(re-frame.core/reg-event-fx
  :process-x
  (fn 
    [{db :db} event-v]
    {:dispatch  ^:flush-dom [:do-work-process-x]   ;; <--- NOW WITH METADATA         
     :db (assoc  db  :processing-X true)}))  ;; so the modal gets rendered

Notice the `^:flush-dom` metadata on the event being dispatched.  Use 
that when you want the UI to be fully updated before the event dispatch 
is handled. 

You only need this technique when you: 

  1. want the DOM to be fully updated
  2. because you are going to hog the CPU for a while and not give it back. One chunk of work.

If you handle via multiple chunks you don't have to do this, because 
you are repeatedly handing back control to the browser/UI.  Its just 
when you are going to tie up the CPU for a one, longish chunk. 
```

-----

The key part for me was `Use that when you want the UI to be fully updated before the event dispatch is handled.` Which I'm assuming is a good thing because we don't want signals dispatching `:signals/signal-received` to immediately interrupt anything, they can wait a tiny bit.

That's why I said the theory is sound, but I couldn't measure differences, but sounds good on paper. Could it have side-effects? I really can't tell.

I can clearly see this technique being useful (again, in theory) in places where we still use `rf/merge` because they hog the CPU for longer in one chunk.

Reply:
Another piece of the docs:

```
;;   - when the event has :flush-dom metadata we schedule via
;;       "reagent.core.after-render"
;;     which will run event processing after the next Reagent animation frame.
```

It sounds good on paper for processing signal events. Or do you have reservations?

Reply:
Now I'm having more questions.

1. Does this metadata work on the `re-frame/dispatch` call or just when we are inside an event handler?
2. I misunderstood and compared it with `js/requestAnimationFrame`, but I think it works the same as `dispatch-sync`, am I right? could we use it instead?
3. > Which I'm assuming is a good thing because we don't want signals dispatching :signals/signal-received to immediately interrupt anything, they can wait a tiny bit.

    So, are you trying to delay the execution of the event? :thinking: if so, I think you don't want to use this approach.
    
I know one or two places in the wallet where this can be used, sometimes we navigate but the re-frame data is not ready yet, so we could dispatch with this metadata so that we don't wait for one frame more. In the past I managed to solve it by delaying the navigation one frame (`js/requestAnimationFrame`), but the team didn't like this approach, you may remember it:
- https://github.com/status-im/status-mobile/pull/18532#issuecomment-1912667235

And the other usage is when we navigate to view a collectible, we store it in re-frame, but it's not ready until the next frame, so we have a one-frame blink

Reply:
Regarding to `rf/merge` yes, I think it can be applied there if there's much work to do

Reply:
I know as much as you about this part @ulisesmac.

I think the implications of this metadata should be explored simply because it's there already for us to be used. It should only make a difference in signals that are handled in the same chunk as `:signals/signal-received`. So for instance, this means the second dispatch to handle a `wallet` signal won't be affected by the metadata optimization, but a signal like `"envelope.sent"` should be affected because it uses `rf/merge`.

> Does this metadata work on the re-frame/dispatch call or just when we are inside an event handler?

I think it works only for the immediate handler. Further re-dispatches from the handler will be treated the same. This metadata is useful in events that are more expensive, as they say in the docs.

> So, are you trying to delay the execution of the event? 

Correct, this is a form of delaying, telling re-frame that the handler for a signal should be processed only after the next Reagent animation frame.

> Does this metadata work on the re-frame/dispatch call or just when we are inside an event handler?

`dispatch-sync` would be the opposite of what I tried in this PR because it would immediately process a signal, without enqueueing, which would more likely affect responsiveness when there are too many (more important) events competing for resources.

> I know one or two places in the wallet where this can be used

It would quite useful if we could experiment with this metadata with real data, as in the wallet, and start to get a feeling of it based on profiling. If it proves to be useless we can always take it away easily.

Reply:
Hey @ilmotta Now I'm not sure if the metadata `:flush-dom` executes  everything "now" or delay the next execution. I personally understand it as a way of `dispatch-sync` but for content that is inside event handlers. So:

> Correct, this is a form of delaying, telling re-frame that the handler for a signal should be processed only after the next Reagent animation frame.

As far as I understand, by default they are executed at the end of the frame, if you are willing to delay it even one frame more, you can wrap it use `js/requestAnimationFrame` or `reagent/next-tick` (they should be exactly the same in React Native).

> dispatch-sync would be the opposite of what I tried in this PR 
What you did in this PR and `dispatch-sync` are, according to what I see, the same thing, not the opposite. But we may need to read the source code to properly understand what is happening.

I've done a test where I know re-frame, sometimes, takes one frame more to propagate the given DB state.
Please consider replicating this example, and it'd be great if it's part of this PR (or should I open a new issue PR for it?)

1. In `status-im.contexts.wallet.collectible.view` move the `collectible` binding in `collectible-details` outside the render function:
    ![image](https://github.com/user-attachments/assets/22906090-f01e-4662-a8a6-478ac55d389e)

2. In `status-im.contexts.wallet.collectible.events` inside the event handler for `:wallet/navigate-to-collectible-details` we navigate to the collectible page delaying the navigation 17ms (slightly more than a frame). So that re-frame propagates its state:
    ![image](https://github.com/user-attachments/assets/1207d2ac-191e-4460-bfd1-31ac4f88eec1)

3. If we use a regular `dispatch`:

    ![image](https://github.com/user-attachments/assets/0136f9f0-edf6-4e07-a7c5-6fed312af8b9)
    
    And with the changes in `1`, once the `collectible` value is defined, it'll never be updated, so if the image is not found and we'll get an error in the UI, here is a video of it, it works as expected:
    
    https://github.com/user-attachments/assets/4b1c1591-689a-45e6-a8d3-723e0f90e13d
    
    As said before, re-frame sometimes is not able to propagate its DB state on time.

4. Now let's just add the metadata `:flush-dom`,
    ![image](https://github.com/user-attachments/assets/35bc2c9f-2137-4915-92f7-0d0b06a81215)
    
    The app now looks smoother and NEVER fails to propagate the data on time:
    
    https://github.com/user-attachments/assets/90ff5d09-4600-4c9f-a46c-e82528c0e9a0
    
So, `:flush-dom`  seems like a way of `dispatch-sync` :thinking: for the event where it's used (not the vent marked with the metadata).


WDYT?

Reply:
Very interesting results @ulisesmac 💯 Happy to see a problem could be solved by the metadata.

But I'm going to remove the metadata from this PR and then we can introduce it with a separate PR focusing on solving a specific problem. I hope that's okay for you :) The thing is that this PR still needs to be QAed, and I'm a bit weary it could introduce a small regression based on our investigations, and since this is maybe the last week before cutting the release branch 🤷🏼 

> So, :flush-dom seems like a way of dispatch-sync 🤔 for the event where it's used (not the vent marked with the metadata).

The implementation of `dispatch-sync` is different, we can see that it literally skips the queue, whereas normal `dispatch` calls `(push event-queue event)`. I prefer to not think of it as dispatch sync for that reason. I saw that there are lots of little comments about the metadata sprinkled in re-frame's source. The comments mention that it can "pause processing". This confused me and it's now unclear to me exactly what the metadata does and I don't trust the docs anymore. For example:

- https://github.com/day8/re-frame/blob/09e2d7132c479aa43f2a64164e54e42bf8511902/src/re_frame/router.cljc#L65
- https://github.com/day8/re-frame/blob/09e2d7132c479aa43f2a64164e54e42bf8511902/src/re_frame/router.cljc#L153
- https://github.com/day8/re-frame/blob/09e2d7132c479aa43f2a64164e54e42bf8511902/src/re_frame/router.cljc#L193

Reply:
> But I'm going to remove the metadata from this PR and then we can introduce it with a separate PR focusing on solving a specific problem. 

Sounds good! @ilmotta 

> The comments mention that it can "pause processing". This confused me and it's now unclear to me exactly what the metadata does and I don't trust the docs anymore. For example:

We can ask in the Clojurians Slack, it's a bit confusing.

Thank you so much for sharing this info :+1:  :+1: